google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStoragePath.java

```diff
@@ -83,8 +86,39 @@ boolean seemsLikeADirectory() {
     return path.seemsLikeADirectory();
   }
 
-  boolean seemsLikeADirectoryAndUsePseudoDirectories() {
-    return path.seemsLikeADirectory() && fileSystem.config().usePseudoDirectories();
+  boolean seemsLikeADirectoryAndUsePseudoDirectories(Storage storage) {
+    if (!fileSystem.config().usePseudoDirectories()) {
+      return false;
+    }
+    if (path.seemsLikeADirectory()) {
+      return true;
+    }
+    // fancy case: the file name doesn't end in slash, but we've been asked to have pseudo dirs.
+    // Let's see if there are any files with this prefix.
+    if (storage == null) {
+      // we are in a context where we don't want to access the storage, so we conservatively
+      // say this isn't a directory.
+      return false;
+    }
+    Page<Blob> list = storage.list(
+        this.bucket(),
+        Storage.BlobListOption.prefix(path.removeBeginningSeparator().toString()),
+        Storage.BlobListOption.pageSize(1));
```

Comment:
What's the rationale for setting this to 1? 

Do we expect this to be resolved in a single iteration? If it's always resolved after a single request and single blob, let's comment that we expect that to be the case.

Reply:
Good point, I added a comment. Yes, we only need to pull a single file.

If for example the user gave us: "foo/bar"

We search for anything that starts with "foo/bar/". If we find anything at all (say, "foo/bar/budget.xls") then we can say that foo/bar is a directory. If there's nothing then "foo/bar" isn't a directory.

I also took the opportunity to add the slash at the end of the search, because I realize otherwise the code might have given the wrong answer.